cmd/tools/handle-release/main.go

```diff
@@ -0,0 +1,377 @@
+			}
+
+			switch {
+			case strings.HasSuffix(key, ".tar.lz4"):
+				a.url = url
+			case strings.HasSuffix(key, ".md5"):
+				a.checksumURL = url
+			case strings.HasSuffix(key, ".txt"):
+				a.packagesURL = url
+			}
+
+			res[base] = a
+		}
+
+		return true
+	})
+	if err != nil {
+		return fmt.Errorf("cannot list s3 objects:%w", err)
+	}
+
+	var artifacts []*artifact
+	for _, a := range res {
+		artifacts = append(artifacts, &a)
+	}
+
+	err = release(artifacts)
+	if err != nil {
+		return err
+	}
+
+	return print(artifacts)
+}
+
+func release(artifacts []*artifact) error {
+	if *dryRun {
+		for _, a := range artifacts {
+			logRunOutput(a)
+		}
+
+		return nil
+	}
+
+	var (
+		errs []error
+	)
+
+	ctx := context.Background()
+	cli, err := docker.NewClientWithOpts(docker.FromEnv, docker.WithAPIVersionNegotiation())
+	if err != nil {
+		return fmt.Errorf("failed to create docker client: %v", err)
+	}
+	defer func() {
+		if err = cli.Close(); err != nil {
+			errs = append(errs, err)
+		}
+	}()
+
+	tok := os.Getenv(token)
+	if tok == "" {
+		return fmt.Errorf("registry token is missing. Please provide TOKEN env variable")
+	}
+
+	var authConfigBase64 string
+	authConfig := registry.AuthConfig{
+		Username:      "metal-stack",
+		Password:      tok,
+		ServerAddress: "ghcr.io",
+	}
+	authConfigBytes, err := json.Marshal(authConfig)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("error encoding authConfig: %v", err))
+		return errors.Join(errs...)
+	}
+	authConfigBase64 = base64.URLEncoding.EncodeToString(authConfigBytes)
+
+	for _, a := range artifacts {
+		logRunOutput(a)
+		sourceImage := a.dockerImage
+
+		pullReader, err := cli.ImagePull(ctx, sourceImage, image.PullOptions{RegistryAuth: authConfigBase64})
+		if err != nil {
+			errs = append(errs, fmt.Errorf("image pull failed: %v", err))
+			return errors.Join(errs...)
+		}
+		defer func() {
+			if err = pullReader.Close(); err != nil {
+				errs = append(errs, err)
+			}
+		}()
+		err = renderDockerOutput(pullReader)
+		if err != nil {
+			errs = append(errs, err)
+			return errors.Join(errs...)
+		}
+
+		for _, t := range a.dockerTags {
+			err = cli.ImageTag(ctx, sourceImage, t)
+			if err != nil {
+				errs = append(errs, fmt.Errorf("image tag failed: %v", err))
+				return errors.Join(errs...)
+			}
+
+			pushReader, err := cli.ImagePush(ctx, t, image.PushOptions{RegistryAuth: authConfigBase64})
+			if err != nil {
+				errs = append(errs, fmt.Errorf("image push failed: %v", err))
+				return errors.Join(errs...)
+			}
+			defer func() {
+				if err = pushReader.Close(); err != nil {
+					errs = append(errs, err)
+				}
+			}()
+			err = renderDockerOutput(pushReader)
+			if err != nil {
+				errs = append(errs, err)
+				return errors.Join(errs...)
+			}
+		}
+
+		fmt.Println()
+
+		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
+		defer cancel()
+
+		client, err := storage.NewClient(ctx)
```

Comment:
Were you able to try this out somehow? Or do I have to do it? If I would do it, I think I would start writing a test first because I can imagine it will be quite hard to maintain this logic. Possible bugs are unfortunately quite devastating in this context and can only be discovered after a release was issued. :(

There is a really cool project [fake-gcs-server](https://github.com/fsouza/fake-gcs-server), which is also used in the backup-restore-sidecar. Maybe it's worth trying it out? Otherwise this would require trial and error and I would need to create some fake buckets and stuff? A test might be more sustainable.

What do you think?